---
title: "A Link To The Past"
subtitle: "Statistical connections within the bioinformatics toolkit"
author: "Russ Hyde"
date: "30th September 2019"
output:
  xaringan::moon_reader
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
half_height = 4
```

## Preamble

Github:

- https://www.github.com/russHyde/bfx_201909

Dataset:

- [GSE103528](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE103528)

R Packages:

```{r, echo = TRUE}
# format: xaringan
pkgs <- c(
  # CRAN
  "magrittr",
  # Bioconductor
  "limma", "edgeR"
)

for (pkg in pkgs) {
  suppressPackageStartupMessages(
    library(pkg, character.only = TRUE))
}
```

---

## The Dataset

[GSE103528](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE103528)

- Sources of variability:

    - 3x AML cell lines (Acute myeloid leukaemia)
    - 3x treatments ("Scramble", and two shRNAs vs _CARM1_ (aka _PRMT4_))
    - Triplicates (not obvious: are replicates batches?)

- Data:

    - Star / RSEM / GRCh38 / Ensembl-87
    - (@russHyde should have aligned the data and used actual counts
        - but life's too short
        - he just `floor`ed the RSEM values)

---

## The Dataset (cont.)

```{r, echo = TRUE, message = FALSE}
dge <- readRDS(file.path("data", "job", "GSE103528.dgelist.rds"))
dge <- dge[dge$genes$gene_biotype == "protein_coding", ]
dim(dge)
```

```{r, echo = TRUE, fig.height = half_height, fig.align = "center"}
dge %>%
  edgeR::cpm(log = TRUE) %>%
  limma::plotDensities(legend = FALSE, main = "Density of log-CPM")
```

---

## The Simplest Model (in shorthand)

$$\mathbf{y} \sim \mathcal{N} \left(X\boldsymbol{\beta}, \sigma^2I \right)$$

Where,

$\mathbf{y}$ - observed values

$\mathtt{lhs} \sim \mathcal{N}(., .)$ - normality assumption (_on_ the
residuals; or _around_ the fitted values)

$X$ - _a_ design matrix (one row per observation)

$\boldsymbol{\beta}, \sigma$ - parameters to be estimated (vector and scalar,
resp)

... plus lots of assumptions

---

## The Simplest Model (in practice)

$$\mathbf{y} \sim \mathcal{N} \left(X\boldsymbol{\beta}, \sigma^2I \right)$$

Where,

$\mathbf{y}$ - eg, log-expression-intensity for a single gene across a set of
samples

$\mathtt{lhs} \sim \mathcal{N}(., .)$ - some statistical distribution

$X$ - some encoding of experimental design / predictors for the samples

$\boldsymbol{\beta}$ - how much does each predictor contribute ...

$\sigma$ - and how much noise is there around ...

... the fitted values

<!-- TODO: Generalisation:
    - Multiple 'y'
      - repeated measures
      - different genes
    - _N_ -> ARBITRARY_DIST(params)
    - Fixed effects -> Mixed effects: modification of X
-->

<!--

## The Simplest Model (Assumptions)

Residuals:

- IID normal

- mean 0 for all values of predictor variables

Model:

- The model is correct ...

- Predictors are fixed and have no errors

- Single source of unaccounted noise

-->

<!--
  - How should the dependent variable be modelled
  - How should the explanatory / independent variables be incorporated
  - What should be estimated
  - What should be controlled
  - How should the covariates be controlled
  - How should uncertainties in mapping, biases in library composition etc etc
  be incorporated
-->

<!--
  Illustrate with a single gene from the example dataset
-->

<!--
  There is _no_ one simplest model
-->

---

## Design Matrices

Encode the experimental / statistical design:

- rows = samples
- columns = model coefficients

ExploreModelMatrix: https://csoneson.github.io/ExploreModelMatrix/

```{r}
design_heatmap <- function(...) {
  heatmap(..., Rowv = NA, Colv = NA, scale = "none",
    xlab = "Samples", ylab = "Coefficients",
    col = c("white", "black"),
    cexCol = 0.5
  )
}
```

.pull-left[
```{r, design-heatmap, eval = FALSE, echo = TRUE, fig.height = half_height}
# design_heatmap plots binary matrices as heatmaps
design_heatmap(
  model.matrix(
    ~ treatment,
    data = dge$samples))
```
]

.pull-right[
```{r, design-heatmap-out, ref.label = "design-heatmap", echo = FALSE}
```
]

---

## The 'transform to normality' approach: `lm()`

---

## Some Univariate Distibutions

```{r, echo = FALSE}
# source: wikipedia
url <- "./figures/Relationships_among_some_of_univariate_probability_distributions.jpg"
```

```{r}
knitr::include_graphics(url)
```

Source: Ehsan Azhdari [CC BY-SA 3.0] via Wikipedia

---

## The 'borrow information' approach: `limma`

---

## The 'count all the things' approach: `edgeR` / `DESeq2`

---

## The Statistical Philosophies

---

## The 'batch-effects are random variables' approach (`lme4`)

<!-- Voom with duplicate correlation -->

<!-- variancePartition::dream() -->

---

## The 'everything is a random variable' approach (`stan` / `jags`)

---

## The 'beyond the scope of the talk, and of my explanatory powers' approaches ...

- genes aren't independent of each other ...

    - epigenomic context

    - coregulatory modules

- dynamic modelling of gene expression ...

- the cell population milkshake ...
