---
title: "Stats & Bioinformatics"
author: "Russ Hyde"
date: "14 August 2019"
output:
    ioslides_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

# A Link To The Past:  Statistical connections within the bioinformatics toolkit


## The Dataset

<!--
  TODO:
  - download some external dataset that has multiple nested levels
  - eg, GSE103528: three treatments, three cell lines, duplicates for each cell
  line
  - Convert it to a count matrix for use in edgeR, limma etc
-->


## The Simplest Model (in shorthand)

$$\mathbf{y} \sim \mathcal{N} \left(X\boldsymbol{\beta}, \sigma^2I \right)$$

Where,

$\mathbf{y}$ - observed values

$\mathtt{lhs} \sim \mathcal{N}(., .)$ - normality assumption (_on_ the
residuals; or _around_ the fitted values)

$X$ - _a_ design matrix (one row per observation)

$\boldsymbol{\beta}, \sigma$ - parameters to be estimated (vector and scalar,
resp)

... plus lots of assumptions


## The Simplest Model (in practice)

$$\mathbf{y} \sim \mathcal{N} \left(X\boldsymbol{\beta}, \sigma^2I \right)$$

Where,

$\mathbf{y}$ - eg, log-expression-intensity for a single gene across a set of
samples

$\mathtt{lhs} \sim \mathcal{N}(., .)$ - some statistical distribution

$X$ - some encoding of experimental design / predictors for the samples

$\boldsymbol{\beta}$ - how much does each predictor contribute ...

$\sigma$ - and how much noise is there around ...

... the fitted values

<!-- TODO: Generalisation:
    - Multiple 'y'
      - repeated measures
      - different genes
    - _N_ -> ARBITRARY_DIST(params)
    - Fixed effects -> Mixed effects: modification of X
-->

<!--

## The Simplest Model (Assumptions)

Residuals:

- IID normal

- mean 0 for all values of predictor variables

Model:

- The model is correct ...

- Predictors are fixed and have no errors

- Single source of unaccounted noise

-->

<!--
  - How should the dependent variable be modelled
  - How should the explanatory / independent variables be incorporated
  - What should be estimated
  - What should be controlled
  - How should the covariates be controlled
  - How should uncertainties in mapping, biases in library composition etc etc
  be incorporated
-->

<!--
  Illustrate with a single gene from the example dataset
-->

<!--
  There is _no_ one simplest model
-->

## Some Univariate Distibutions

```{r, echo = FALSE}
# source: wikipedia
url <- "./figures/Relationships_among_some_of_univariate_probability_distributions.jpg"
```

![](`r url`){ width=100% }

Source: Ehsan Azhdari [CC BY-SA 3.0] via Wikipedia

## The 'transform to normality' approach: `lm()`

## The 'borrow information' approach: `limma`

## The 'count all the things' approach: `edgeR` / `DESeq2`

## The Statistical Philosophies

## The 'batch-effects are random variables' approach (`lme4`)

<!-- Voom with duplicate correlation -->

<!-- variancePartition::dream() -->

## The 'everything is a random variable' approach (`stan` / `jags`)

## The 'beyond the scope of the talk, and of my explanatory powers' approaches ...

- genes aren't independent of each other ...

    - epigenomic context

    - coregulatory modules

- dynamic modelling of gene expression ...

- the cell population milkshake ...
