---
title: "A Link To The Past"
subtitle: "Statistical connections within the bioinformatics toolkit"
author: "Russ Hyde"
date: "30th September 2019"
output:
  xaringan::moon_reader
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
height.5 = 6
height.75 = 9
set.seed(3)
```

```{r}
design_heatmap <- function(
  ...,
  row_title = "Samples",
  column_title = "Coefficients"
) {
  ComplexHeatmap::Heatmap(...,
    col = c("0" = "grey87", "1" = "black"),
    cluster_rows = FALSE,
    cluster_columns = FALSE,
    row_title = row_title,
    column_title = column_title
  )
}
```

```{r}
factor_model <- function(
  object, ...
) {
  stopifnot(is(object, "formula"))

  mm <- stats::model.matrix(object, ...)

  # prefixing the factor levels with the factor name seems odd

  term_names <- paste0("^", attr(terms(object), "term.labels"))

  colnames(mm) <- stringr::str_replace_all(
    colnames(mm),
    pattern = magrittr::set_names(rep("", length(term_names)), term_names)
  )

  mm
  }
```

## Preamble

Github:

- https://www.github.com/russHyde/bfx_201909

Dataset:

- [GSE103528](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE103528)

R Packages:

```{r, echo = TRUE}
# format: xaringan
pkgs <- c(
  # CRAN
  "magrittr",
  # Bioconductor
  "limma", "edgeR", "ComplexHeatmap"
)

for (pkg in pkgs) {
  suppressPackageStartupMessages(
    library(pkg, character.only = TRUE))
}
```

---

## The Dataset

[GSE103528](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE103528)

- Sources of variability:

    - 3x AML cell lines (Acute myeloid leukaemia)
    - 3x treatments ("Scramble", and two shRNAs vs _CARM1_ (aka _PRMT4_))
    - Triplicates (not obvious: are replicates batches?)

- Data:

    - Star / RSEM / GRCh38 / Ensembl-87
    - (@russHyde should have aligned the data and used actual counts
        - but life's too short
        - he just `floor`ed the RSEM values)

---

## The Dataset (cont.)

```{r, echo = TRUE, message = FALSE}
dge <- readRDS(file.path("data", "job", "GSE103528.dgelist.rds"))
dge <- dge[dge$genes$gene_biotype == "protein_coding", ]
dim(dge)
```

.pull-left[

```{r density, eval = FALSE, echo = TRUE}
dge %>%
  edgeR::cpm(log = TRUE) %>%
  limma::plotDensities(
    legend = FALSE,
    main = "Density of log-CPM"
    )
```
]

.pull-right[
```{r, density-out, ref.label = "density", echo = FALSE, fig.height = height.5}
```
]

---

## The Simplest Model (in shorthand)

$$\mathbf{y} \sim \mathcal{N} \left(X\boldsymbol{\beta}, \sigma^2I \right)$$

Where,

$\mathbf{y}$ - observed values

$\mathtt{lhs} \sim \mathcal{N}(., .)$ - normality assumption (_on_ the
residuals; or _around_ the fitted values)

$X$ - _a_ design matrix (one row per observation)

$\boldsymbol{\beta}, \sigma$ - parameters to be estimated (vector and scalar,
resp)

... plus lots of assumptions

---

## The Simplest Model (in practice)

$$\mathbf{y} \sim \mathcal{N} \left(X\boldsymbol{\beta}, \sigma^2I \right)$$

Where,

$\mathbf{y}$ - eg, log-expression-intensity for a single gene across a set of
samples

$\mathtt{lhs} \sim \mathcal{N}(., .)$ - some statistical distribution

$X$ - some encoding of experimental design / predictors for the samples

$\boldsymbol{\beta}$ - how much does each predictor contribute ...

$\sigma$ - and how much noise is there around ...

... the fitted values

<!-- TODO: Generalisation:
    - Multiple 'y'
      - repeated measures
      - different genes
    - _N_ -> ARBITRARY_DIST(params)
    - Fixed effects -> Mixed effects: modification of X
-->

<!--

## The Simplest Model (Assumptions)

Residuals:

- IID normal

- mean 0 for all values of predictor variables

Model:

- The model is correct ...

- Predictors are fixed and have no errors

- Single source of unaccounted noise

-->

<!--
  - How should the dependent variable be modelled
  - How should the explanatory / independent variables be incorporated
  - What should be estimated
  - What should be controlled
  - How should the covariates be controlled
  - How should uncertainties in mapping, biases in library composition etc etc
  be incorporated
-->

<!--
  Illustrate with a single gene from the example dataset
-->

<!--
  There is _no_ one simplest model
-->

---

## Design Matrices

.pull-left[
Encode the experimental / statistical design:

- rows = samples
- columns = model coefficients

```{r, design-heatmap, eval = FALSE, echo = TRUE}
# `design_heatmap` plots binary matrices as heatmaps
# `factor_model` wraps `model.matrix` - but doesn't prefix with the factor name
design_heatmap(
  factor_model(
    ~ cell + treatment,
    data = dge$samples)
) +
  rowAnnotation(
    df = dge$samples[, c("cell", "batch", "treatment")]
  )
```

See also: ExploreModelMatrix

- https://csoneson.github.io/ExploreModelMatrix/
]

.pull-right[
```{r, design-heatmap-out, ref.label = "design-heatmap", echo = FALSE, fig.height = height.75}
```
]

---

## The 'transform to _normality_' approach: `lm()`

```{r}
design <- factor_model(~ cell + treatment, data = dge$samples)
dge_voom <- voom(dge, design = design)

```

---

## Equivalent Design Matrices

```{r}
hm1 <- design_heatmap(
  factor_model(~ cell + treatment, data = dge$samples),
  column_title = "~ cell + treatment[baseline:sh55]",
  name = "design_1"
)
```

```{r}
ra <- rowAnnotation(df = dge$samples[, c("cell", "batch", "treatment")])
```

```{r}
hm2 <- design_heatmap(
  factor_model(
    ~ -1 + cell + treatment,
    data = dplyr::mutate(
      dge$samples, treatment = relevel(factor(treatment), "shct"))
    ),
  column_title = "~ -1 + cell + treatment[baseline:shct]",
  name = "design_2"
)
```

```{r, fig.width = 10}
draw(hm1 + ra + hm2, ht_gap = unit(3, "cm"))

f <- function(i0) {
  i <- i0
  y <- c(i0 - 1, i0) / nrow(dge$samples)
  grid.lines(c(0, 1), y[1], gp = gpar(col = "red"))
  grid.lines(c(0, 1), y[2], gp = gpar(col = "red"))
}

decorate_heatmap_body("design_1", f(7))
decorate_heatmap_body("design_2", f(7))
```

---

## The 'borrow information' approach: `limma`

---

## Some Univariate Distibutions

```{r, echo = FALSE}
# source: wikipedia
url <- "./figures/Relationships_among_some_of_univariate_probability_distributions.jpg"
```

```{r}
knitr::include_graphics(url)
```

Source: Ehsan Azhdari [CC BY-SA 3.0] via Wikipedia

---

## The 'count all the things' approach: `edgeR` / `DESeq2`

---

## The Statistical Philosophies

---

## The 'batch-effects are random variables' approach (`lme4`)

<!-- Voom with duplicate correlation -->

<!-- variancePartition::dream() -->

---

## The 'everything is a random variable' approach (`stan` / `jags`)

---

## The 'beyond the scope of the talk, and of my explanatory powers' approaches ...

- genes aren't independent of each other ...

    - epigenomic context

    - coregulatory modules

- dynamic modelling of gene expression ...

- the cell population milkshake ...
